armsrc/lfsampling.c

```diff
@@ -424,6 +425,107 @@ uint32_t SampleLF(bool verbose, uint32_t sample_size, bool ledcontrol) {
+    BigBuf_Clear_ext(false);
+    LFSetupFPGAForADC(config.divisor, reader_field);
+
+    // parameters from config and constants
+    const uint8_t bits_per_sample = config.bits_per_sample;
+    const int16_t trigger_threshold = config.trigger_threshold;
+    int32_t samples_to_skip = config.samples_to_skip;
+    const uint8_t decimation = config.decimation;
+
+    uint32_t sample_size = 64;
+    const int8_t size_threshold_table[9] = {0, 64, 64, 48, 64, 40, 48, 56, 64};
+    const int8_t size_threshold = size_threshold_table[decimation];
+
+    // DoAcquisition() start
+    uint8_t last_byte = 0;
+    uint8_t curr_byte = 0;
+    int return_value = PM3_SUCCESS;
+
+    initSampleBuffer(&sample_size); // sample size in bytes
+    sample_size <<= 3; // sample size in bits
+    sample_size /= bits_per_sample; // sample count
+
+    bool trigger_hit = false;
+    int16_t checked = 0;
+
+    while (BUTTON_PRESS() == false) {
+
+        // only every 4000th times, in order to save time when collecting samples.
+        // interruptible only when logging not yet triggered
+        if ((checked >= 4000) && trigger_hit == false) {
+            if (data_available()) {
+                checked = -1;
+                break;
+            } else {
+                checked = 0;
+            }
+        }
+        ++checked;
+
+        WDT_HIT();
+
+        if ((AT91C_BASE_SSC->SSC_SR & AT91C_SSC_TXRDY)) {
+            LED_D_ON();
+        }
+
+        if (AT91C_BASE_SSC->SSC_SR & AT91C_SSC_RXRDY) {
+            volatile uint8_t sample = (uint8_t)AT91C_BASE_SSC->SSC_RHR;
+
+            // Test point 8 (TP8) can be used to trigger oscilloscope
+            LED_D_OFF();
+
+            // threshold either high or low values 128 = center 0.  if trigger = 178
+            if (trigger_hit == false) {
+                if ((trigger_threshold > 0) && (sample < (trigger_threshold + 128)) && (sample > (128 - trigger_threshold))) {
+                    continue;
+                }
+            }
+
+            trigger_hit = true;
+
+            if (samples_to_skip > 0) {
+                samples_to_skip--;
+                continue;
+            }
+
+            logSample(sample, decimation, bits_per_sample, false);
+
+            // write to USB FIFO if byte changed
+            curr_byte = data.numbits >> 3;
+            if (curr_byte > last_byte)
+                usb_write_byte_async(data.buffer[last_byte]);
+            last_byte = curr_byte;
+
+            if(samples.total_saved == size_threshold)
+            {
+                // request usb transmission and change FIFO bank
+                if (usb_write_request() == false) {
+                    return_value = PM3_EIO;
+                    break;
+                }
+
+                // reset sample
+                last_byte = 0;
+                data.numbits = 0;
+                // data.position = 0; // unused?
+                samples.total_saved = 0;
+            }
+        }
+    }
+
```

Comment:
After button is pressed, or the loop exist for another reason, could there be one or more samples that were pushed into the hardware FIFO buffer?  If so, should a final call to `usb_write_request()` occur?
Careful ... also have to consider error exit paths ... which may indicate those final bytes are better off lost?

Reply:
I haven't implement this part yet. I need to design the overall behavior when the user-required exit or error occurs for both the client and the firmware, so I might do it when the basic sampling code for client and the firmware is finished.

Reply:
I've finished that. Please check commit 079d2c7e37e116e82010715185e56c13547feb72